doc/readthedocs/codingguidelines.rst

```diff
@@ -6,33 +6,85 @@ Coding Guidelines
-------------------------------------
+If necessary, use smart pointers for their memory management capabilities, e.g., ``std::shared_ptr``, ``std::unique_ptr``
+(you may find more information on passing smart pointers
+`here <https://www.modernescpp.com/index.php/c-core-guidelines-passing-smart-pointer/>`_
 
-**Preamble:** |FOURC| is a legacy code.
-Large parts were written pre-C++11 and are not modern C++.
-In addition, code review was not widely practiced back in these days.
-Thus, it is recommended to critically examine the code and techniques while working on the code base and refer to the general guidelines above.
+By default, pass parameters by const reference and only expose smart pointers if memory management is necessary.
+
+See also `<https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rr-owner>`_.
+
+Use of ``Teuchos::ParameterList``
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Prefer parameter container classes over ``Teuchos::ParameterList`` for passing parameters to functions. Often,
+a simple struct with all necessary data as public, non-const members is sufficient.
+
+Reason: ``Teuchos::ParameterList`` is a flexible container for input(!) parameters, but it is also very slow and heavy.
+In addition, the string-based lookup is not compile-time checked.
+
+Const-correctness
+^^^^^^^^^^^^^^^^^
+
+Make new code const-correct and fix old code in that regard when working on it. Const-correctness is mainly relevant
+for function parameters and member functions.
+
+**Note:** The member fields of a struct or class should often not be ``const``, as this would prevent the struct or
+class from being moved or copied. Instead, use ``const`` member functions to access the fields in a ``const`` context.
+
+Enums
+^^^^^
 
-The following guidelines are especially relevant for the current state of |FOURC|:
+- By default, use scoped ``enum class`` instead of unscoped ``enum``.
+- Use enums instead of strings for parameters that take one from a fixed set of values.
+- You do not have to write conversion functions between enum constants and strings.
+  Use ``magic_enum::enum_name`` to get the name of an enum constant and ``magic_enum::enum_cast`` to convert a string to
+  an enum constant.
 
-- If necessary, use smart pointers for their memory management capabilities, e.g., ``std::shared_ptr``, ``std::unique_ptr``
-  (you may find more information on passing smart pointers `here <https://www.modernescpp.com/index.php/c-core-guidelines-passing-smart-pointer/>`_
-- by default, pass parameters by const reference and only expose smart pointers if memory management is necessary
-- Prefer parameter container classes over ``Teuchos::ParameterList`` for passing parameters among element routines
-- Make new code const-correct and fix old code in that regard when working on it.
-- Implement new features in the new structural time integration. Work towards migrating existing capabilities to the new structural time integration.
```

Comment:
Not 100% sure here, but it seems this does not fit in with the other guidelines.